### ISSUE TITLE
[RFC] construct-specific bridgebuilder personas + taste loop + cognition layer

### ISSUE THREAD

**Issue by zkSoju** [0xHoneyJar/loa#310]
## summary

bridgebuilder persona enrichment (loa#295, PR #301) delivered the infrastructure: `persona-loader.ts`, injection detection, 5 persona files, enriched review workflow. the next step is **construct-specific personas that activate domain expertise during review**, and a **taste loop** where bridgebuilder critiques feed back into construct refinement.

additionally: a new persona type — the **cognition layer persona** — that doesn't just review code but **discovers the questions we're afraid to ask**.

## what exists

from PR #301 (v1.35.1):
- `persona-loader.ts` loads persona files with injection detection (14 patterns)
- persona files at `.claude/skills/bridgebuilder-review/resources/personas/`:
  - `default.md` — quality/correctness
  - `security.md` — security audit
  - `taste.md` — artisan aesthetic critique
  - `research.md` — observer research critique
  - `strategy.md` — GTM critique
- `persona_path` config in `.loa.config.yaml`
- YAML frontmatter in persona files for model selection

### 1. pack-to-persona mapping incomplete

5 packs installed (observer, artisan, beacon, crucible, gtm), but only 2 have matching personas (artisan→taste, observer→research). beacon and crucible have no review persona. the manifest should declare which persona to use:

```yaml
# in manifest.json
bridgebuilder:
  persona: "research"        # maps to personas/research.md
  persona_override: "grimoires/observer/bridgebuilder-persona.md"  # optional custom
```

### 2. taste loop not implemented

the backward pass of the neural network metaphor: bridgebuilder critiques should feed back into construct persona refinement.

```
construct output → bridgebuilder review (with domain persona)
                        ↓
               critique findings
                        ↓
               propose persona.md refinements
                        ↓
               user approves → persona updated
```

this is how the system "learns" — each review cycle refines the construct's persona, which refines future output.

### 3. cognition layer persona (new type)

from midi-interface#54:

> "the persona should grill me as if I don't know what would hurt the most to hear but would expose the most truth about the user that could inform the product. the diamond in the rough, the gems in the sand, dirt turned gold."

this isn't a code review persona. it's a **cognition layer** — system 2 thinking applied to user research. it should:
- read canvases and score API data
- identify what we're afraid to learn about each user
- output the 3 questions that would hurt most to hear the answers to, but would expose the most product truth
- act as a pre-follow-up step: discover scary questions → generate messages that steer there

this persona doesn't review code. it reviews our **understanding of the user** and finds the holes.

### 4. persona invocation per construct via bridgebuilder

the user wants to invoke bridgebuilder with a specific persona for a specific construct's output:

```
/run-bridge --persona research --construct observer
```

this would:
1. load the observer construct's context (base + overlays)
2. load the research persona
3. run bridge review with both loaded — domain expertise + review methodology
4. output critiques specific to user research quality, not code quality

## connection to hounfour

hounfour's `persona-loader.ts` already supports this pattern. the `ensemble.ts` module supports running multiple model-backed reviews with `consensus` merge. the missing adapter is:

1. construct manifest declares persona preference
2. CLI reads manifest → translates to hounfour `AgentBinding`
3. hounfour routes to appropriate model with persona loaded
4. review output feeds back into taste loop

## related issues

- loa#295 — bridgebuilder persona enrichment (completed, PR #301)
- loa-finn#29 — beauvoir pattern transfer for bridgebuilder
- loa-finn#31 — main hounfour RFC
- midi-interface#54 — observer golden path + construct bridgebuilder persona
- midi-interface#53 — follow-up messages should probe scoring mental model

## the pseudo-training metaphor

> "in my head in training your own model (even though in this case we don't use gpus, we're pseudo training with prompts and weights)... the mixture of agents stands out to me bc it's connected to how I would approach the fullstack work required"

each review cycle is a training step:
- **forward pass**: construct generates output (follow-ups, canvases, syntheses)
- **loss function**: bridgebuilder persona reviews and scores
- **backward pass** (taste loop): critique findings refine the persona
- **weight update**: updated persona changes future construct behavior

the constructs ARE the model weights. the personas ARE the loss functions. the taste loop IS backpropagation.

**Comment by zkSoju** [comment]
we need to be aware here that there is a clear distinct on the purpose of Loa Runtime vs the Constructs Network and that the bridgebuilder config provided by the runtime would like to be consumed by the constructs in the constructs network. i would even go deeper to say that this could be dialed in further to a skill by skill basis. tbh this is already implemented sort've here and scattered into skills so wondering if we hoist it all into bridgebuilder it would be cleaner to work with. `.claude/prompts/gpt-review/base`

**Comment by janitooor** [comment]
**Status update (v1.66.0)**:

cycle-051 delivers the infrastructure this RFC depends on:
- Construct index includes `persona_path` per construct — the Bridgebuilder can now discover construct personas programmatically
- Name resolution maps construct names to their full metadata including persona
- `construct-resolve.sh capabilities <slug>` returns construct capability profile

**What remains from this RFC:**
- Bridgebuilder integration: using construct-specific personas during review (not just the default BEAUVOIR.md)
- Taste loop: learning which persona produces the best reviews per construct domain
- Cognition layer: the deeper architectural proposal

This RFC is still relevant for its Bridgebuilder-specific proposals. The infrastructure dependency (construct index + persona discovery) is now resolved.

**Comment by deep-name** [comment]
## Status Update (cycle-082, 2026-04-16)

**Partially implemented.** Current state of the three sub-goals:

### 1. Construct-specific personas ✅ Partial
- 5 built-in personas exist in `.claude/data/`: default, security, dx, architecture, quick
- Persona path configurable per-project via `.loa.config.yaml`
- **Gap:** No pack manifest → persona mapping. Constructs can't declare their preferred reviewer persona.

### 2. Taste loop ❌ Not implemented
- No feedback mechanism where bridge review outcomes influence future persona selection
- Vision entries from bridge reviews are captured but not fed back into persona weighting

### 3. Cognition layer ❌ Not implemented
- No meta-cognitive layer that adapts review depth/focus based on project maturity
- The concept of "review personality that evolves with the codebase" remains aspirational

**Narrowed scope:** The persona infrastructure is in place. This issue now tracks:
- Pack manifest `persona` field in construct schema (ties a construct to its preferred reviewer)
- Taste loop: bridge finding patterns → persona weight adjustments
- Cognition layer: project maturity signals → review depth calibration

**Comment by deep-name** [comment]
## Tangentially addressed by #396 (closed via PR #588, v1.98.0)

The persona DX portion of this RFC's scope — specifically the \"how do operators discover and debug persona resolution\" question — was addressed in [PR #588](https://github.com/0xHoneyJar/loa/pull/588) / issue #396:

- \`--list-personas\` CLI flag enumerates the 5 built-in packs with H1 titles
- \`--show-persona-resolution\` traces the 5-level cascade showing which level wins, which are shadowed, and which are missing (the previously-silent Level 4 fallthrough is now marked \`[missing]\` with an actionable reason)

This is the *discovery + debuggability* surface, not the *construct-specific activation* that this RFC's broader scope covers.

## What this RFC still covers (untouched by #396)

1. **Construct-specific persona activation** — a persona tied to a construct activates when the construct is mounted, providing domain expertise during review
2. **Taste loop** — bridgebuilder critiques feed back into construct refinement
3. **Cognition-layer persona** — reviews that \"discover the questions we're afraid to ask\" rather than just checking code
4. **Inter-construct persona inheritance** — cross-construct persona composition

These remain valid and valuable as a separate body of work.
